diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,217 @@
+        distribution: String,
+        version: String,
+        data_directory: PathBuf,
+        monitoring_port: u16,
+        is_monitoring_enabled: bool,
+        is_reporting_enabled: bool,
+    ) -> Self {
+        let diagnostics = Arc::new(Diagnostics::new(
+            deployment_id.clone(),
+            server_id,
+            distribution,
+            version,
+            data_directory.clone(),
+            is_reporting_enabled,
+        ));
+        let reporter =
+            Reporter::new(deployment_id, diagnostics.clone(), REPORTING_URI, data_directory, is_reporting_enabled);
+        let monitoring_server = MonitoringServer::new(diagnostics.clone(), monitoring_port, is_monitoring_enabled);
+
+        Self { diagnostics, reporter, monitoring_server }
+    }
+
+    diagnostics_method! {
+        pub fn submit_database_metrics(&self, database_metrics: HashSet<DatabaseMetrics>);
+        pub fn submit_error(&self, database_name: Option<&str>, error_code: String);
+        pub fn submit_action_success(&self, database_name: Option<&str>, action_kind: ActionKind);
+        pub fn submit_action_fail(&self, database_name: Option<&str>, action_kind: ActionKind);
+        pub fn increment_load_count(&self, database_name: &str, connection_: LoadKind);
+        pub fn decrement_load_count(&self, database_name: &str, connection_: LoadKind);
+    }
+
+    pub fn may_start_reporting(&self) {
+        self.reporter.may_start()
+    }
+
+    pub fn may_start_monitoring(&self) {
+        self.monitoring_server.may_start()
+    }
+}
+
+pub struct DiagnosticsResultTracker<'d> {
+    diagnostics_manager: &'d DiagnosticsManager,
+    database_name: Option<&'d str>,
+    action_kind: ActionKind,
+    error_code: Option<String>,
+}
+
+impl<'d> DiagnosticsResultTracker<'d> {
+    pub fn without_database(diagnostics_manager: &'d DiagnosticsManager, action_kind: ActionKind) -> Self {
+        DiagnosticsResultTracker { diagnostics_manager, database_name: None, action_kind, error_code: None }
+    }
+
+    pub fn with_database(
+        diagnostics_manager: &'d DiagnosticsManager,
+        database_name: &'d str,
+        action_kind: ActionKind,
+    ) -> Self {
+        DiagnosticsResultTracker {
+            diagnostics_manager,
+            database_name: Some(database_name),
+            action_kind,
+            error_code: None,
+        }
+    }
+
+    pub fn set_error_code(&mut self, error_code: String) {
+        self.error_code = Some(error_code);
+    }
+}
+
+impl<'d> Drop for DiagnosticsResultTracker<'d> {
+    fn drop(&mut self) {
+        match &self.error_code {
+            Some(error_code) => {
+                self.diagnostics_manager.submit_action_fail(self.database_name, self.action_kind);
+                self.diagnostics_manager.submit_error(self.database_name, error_code.clone());
+            }
+            None => self.diagnostics_manager.submit_action_success(self.database_name, self.action_kind),
+        }
+    }
+}
```

Comment:
An interesting approach.

Reply:
In a positive, neutral, or negative way? =)

Reply:
I was undecided until I saw where it was used :P